Refactor to remove usage of post related effects in packages/editor.

nerrad opened the pull request:
## Description
<!-- Please describe what you have changed or added -->

This pull is the first step in moving away from the lingering usage of effects in various data stores among packages.  This pull specifically will deal with the `@wordpress/editor` package (`core/editor` store).

For the most part the following goals are in mind:

- organize file structure so things are a bit more clearly sorted.
- As much as possible I'm preserving the existing logic.  This means shifting up an action creator that returns a action.type that had effects into an action generator (i.e. `savePost`) and then moving the effects into the action-generator. Dispatches and selects in the effects would then be moved into the generator and yield _control_ dispatch and selects (same for usage apiFetch etc).
- Of course refactoring tests for all the changes.
- This will also allow me to keep things in the existing files until the last moment  when I'm ready to roll over to using the new code and test.

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

* [x] covered changes in unit tests
* [x] existing e2e tests coverage
* [x] Tested various post saves and status changes via browser testing to ensure there's no issues.

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

There are no breaking changes in this pull.  The interfaces remain effectively the same. However:

- See the note below regarding the trashPost action and some consistency introduction there. The signature has changed for `trashPost` but ultimately that does not affect any behaviour.
- Some new actions are exposed to retain existing behaviour from the effects.  These actions are not really something that should be publicly exposed so they are prefixed with `__experimental` so that provides room for later iterations potentially changing/removing them.  I didn't do any refactoring there because the goal of this pull is to just convert from effects to controls/action-generators.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://wordpress.org/gutenberg/handbook/designers-developers/ -->

## Additional Notes

### `trashPost` action.

Currently `trashPost` accepts a `postId` and a `postTypeSlug` but they are inherently ignored by the effect callback handling the actual trash in favor of trashing the currentPost.  I [brought this up in slack](https://wordpress.slack.com/archives/C02QB2JS7/p1549590206300100) and @aduth and I resolved that:

- `core/editor` can keep `trashPost` but drop the args.  The interface then assumes that `trashPost` means trash the current post in the state for the editor.
- at some point in the future we could introduce a more explicit `trashPost( postId, postType )` for the `core` store that would trash the given post for the given id.

Related #11124

nerrad committed:
add controls for fetch, select, and dispatch

nerrad committed:
move _all_ constants into the constants file for better organization

nerrad committed:
establish file structure, begin refactoring post effects/actions

nerrad commented:
Need to implement `createRegistryControl` from #13722 when it lands in master

nerrad committed:
Implement trashPost action-generator

- also rename functions for getting notification dispatch arguments (a little less mouthy)

nerrad commented:
Effects to be refactored:

* [ ] REQUEST_POST_UPDATE
* [ ] REQUEST_POST_UPDATE_SUCCESS
* [ ] REQUEST_POST_UPDATE_FAILURE
* [ ] TRASH_POST
* [ ] TRASH_POST_FAILURE
* [ ] REFRESH_POST
* [ ] MERGE_BLOCKS
* [ ] SETUP_EDITOR
* [ ] RESET_BLOCKS
* [ ] SYNCHRONIZE_TEMPLATE
* [ ] FETCH_REUSABLE_BLOCKS
* [ ] SAVE_REUSABLE_BLOCK
* [ ] DELETE_REUSABLE_BLOCK
* [ ] RECEIVE_REUSABLE_BLOCKS
* [ ] CONVERT_BLOCK_TO_STATIC
* [ ] CONVERT_BLOCK_TO_REUSABLE
* [ ] REMOVE_BLOCKS
* [ ] REPLACE_BLOCKS
* [ ] MULTI_SELECT

youknowriad commented:
Do you think it's possible to ship PRs for these effects one after the other?

nerrad commented:
I'm wary about doing multiple pulls because I'm not just refactoring I'm also re-organizing.  That's a lot of potential for merge conflicts.  With that said, what I could do is ship all the post related effect refactors first and then do the block related affects in a different pull.  Would something like that work better?

nerrad committed:
fix import jsdocs

youknowriad commented:
> Would something like that work better?

Yes, If it's a big PR, it's very hard to review properly. splitting is better.

nerrad committed:
update controls to use new createRegistryControl interface

nerrad committed:
add action generator for refreshing the current post

nerrad committed:
add tests for new code

- also removes old code that’s been replaced and fixes up imports

nerrad committed:
fix broken tests

nerrad commented:
Getting close to done.  However there are some failing e2e tests I need to look at (initial look shows it appears to be something involving autopost saves).

Browser testing has gone okay for me so far.  If anyone wants to start reviewing this that would be fine.  I'll keep looking into why the tests are failing.

nerrad commented:
So browser testing is working great for me.  I _think_ the issue with the e2e tests is caused by `post.id` not set in the `savePost` action triggered by autosave in the e2e tests.  It's possible there's some special autosave handling in the e2e tests that is affected by this refactor (haven't dug in yet).  Just making these notes for myself so I don't forget.

nerrad committed:
don’t continue with refreshPost if there’s no current post yet

nerrad commented:
> So browser testing is working great for me.

Sigh, I didn't have the GB plugin active :/  I'm reproducing issues while user testing so clearly something is borked.

nerrad commented:
whoah, found the cause.  The issue is `redux-routine` determines if something "quacks like an action" by the presence of a `type` property.  The selector for `getCurrentPost` is return a post object which has a type property.  So `redux-routine` considers that another action to dispatch :/

nerrad committed:
remove unnecessary bail

aduth commented:
> whoah, found the cause. The issue is `redux-routine` determines if something "quacks like an action" by the presence of a `type` property. The selector for `getCurrentPost` is return a post object which has a type property. So `redux-routine` considers that another action to dispatch :/

Related Slack conversation ([link requires registration](https://make.wordpress.org/chat/)):

https://wordpress.slack.com/archives/C02QB2JS7/p1549904073023400

nerrad commented:
I've tested the change in https://github.com/WordPress/gutenberg/pull/13818 in this branch and it resolves the issues I was seeing here.  So once that pull lands and we get it in this branch this should be good to go to (after review of course)

nerrad committed:
add changelog entry

nerrad committed:
fix lint failure for import order

nerrad commented:
rats. still some e2e fails (and linting).  I'm currently debugging the e2e fails locally.

nerrad committed:
fix incorrect method used for autosave requests

nerrad committed:
remove another wayward period.

nerrad commented:
Okay there is only one remaining failing test and it is:

> The command "npm run check-local-changes" exited with 1.

I'm not sure how to resolve this.  I imagine it has to do with a package-lock change when travis runs npm build.  So does a package-lock change need to be generated in a commit? Should it happen in a specific environment?

aduth commented on packages/editor/src/store/actions/index.js:
```diff
@@ -0,0 +1,2 @@
+export * from './post-actions';
```
Is there a compelling reason to segregate the two?

nerrad replied:
Do you mean among the post actions?  or do you mean specifically having "post" type actions and eventually "block" type actions?

If the former, I find it easier to know where to go for general actions as opposed to action generators.  It also aids with splitting up tests to cover the specific things.

aduth replied:
> I find it easier to know where to go for general actions as opposed to action generators.

It doesn't seem to me that it's something someone should need to distinguish. I think a current reality is that we have spaghetti generators, but they're all just "actions (dispatchers)", ideally returning or yielding simple action objects.

nerrad replied:
I'm probably ignorant of potential reasons why we wouldn't want to split up things in this case.  Are there technical/performance/build reasons why its good to avoid splitting up files like this?

nerrad replied:
I usually try to logically split files up into smaller components when I can because:

a. it helps with finding things quicker when I'm just looking for a "class" of logic
b. it makes for smaller number of lines in a file, so less scrolling.
c. My IDE built-in linting etc doesn't get bogged down as much due to larger files.

But as I already mentioned, I'm aware I may be ignorant of good reasons for not splitting things up as I did here.

aduth replied:
To me, it's not a technical point, it's of discoverability and maintainability generally. What we're doing here is creating a subclass (of actions) where otherwise we've aligned to a common pattern of identifying a few top-level behaviors associated with a store: mutations (`actions.js`), state (`reducer.js`), derivations / getters (`selectors.js`), and side-effects (`effects.js`).

I'm not saying this can't be expanded, but we should do so carefully, deliberately, and consistently. Small files are not inherently more maintainable than a single large file, if it's not obvious or predictable why the files are split the way they are. I find the `effects/` folder pattern to be a good example of this, where I think it was better off in single file form, because the current split is not obvious, and in the majority of cases I arrive at the wrong place when trying to track down effects-related behavior.

My prior point was that the division between generator-actions and non-generator-actions is a bit weak, and arguably intentionally so (my point about "you shouldn't need to distinguish"), so creating subclasses within actions for these does not seem an obvious thing to do.

nerrad replied:
Kinda funny that we're both hinging our argument on discoverability and maintainability :).  I'm not a project maintainer so I don't want to break any established conventions for the project.  I'm happy to just combine them into one file (and the same with tests) if desired.  So just for clarification, is this a _directive_ I need to follow in order to get this pull mergeable or is this a preference you are arguing for but is not a directive yet?

To be clear, I did loosely swing my intentions by @youknowriad a bit in dm with him before embarking, although he suggested this structure (which mirrors the current effects):

> Personally, I'd start with adding those to `actions.js` but maybe split it by domains
>
> /actions/index.js
> /actions/post.js
> /actions/reusable-blocks.js

So I _did_ expand on that a bit in my implementation.  So in your clarification, can you also give direction for whether I should just put everything into the existing `actions.js` file (and subsequently add all the tests to the existing `tests/actions.js` as well?

aduth replied:
Nope, no directives. Just surfacing my own opinions stemming from the initial concern / probe for rationale of the split. It seems there's at least a second opinion in your favor from @youknowriad , but I'd go on record to disagree with the "domain"-based split (it wasn't but an hour after my previous complaint about the effects structuring that I wrongly navigated to `effects/posts.js` as my first destination in pursuit of [this snippet of code](https://github.com/WordPress/gutenberg/blob/a99c0b5487ff3639c05a7ce1dbf7823b821382f8/packages/editor/src/store/effects.js#L206-L216)).

To your question, yes, I'd see it all just being in a single `actions.js` file.

There's an interesting overlap between this discussion and some older ones like in #3030, where I think ultimately it was the introduction of `@wordpress/data` and its store-namespacing that served as the alternative to this problem of files becoming large and unwieldy, where the separation was by _stores_ (modules), not by separate files within a store.

youknowriad replied:
`actions.js` is better if we had one file but if the file grows indefinitely, it becomes hard to navigate the file. maybe there's a middleground to find at some point.

I also said to @nerrad that in general I'm not very opinionated on these matters :)

nerrad replied:
> but if the file grows indefinitely, it becomes hard to navigate the file.

Ya this is one criteria I use when deciding whether to split up a file.  For example _just_ with the post actions here, combined, they would be 487 lines.  That's _before_ adding in all the block and editor related actions.

> as my first destination in pursuit of [this snippet of code](https://github.com/WordPress/gutenberg/blob/a99c0b5487ff3639c05a7ce1dbf7823b821382f8/packages/editor/src/store/effects.js#L206-L216)).

@aduth my _plan_ was to have `editor-actions` and `block-actions` files that would accompany the `post-actions`.  That snippet is found within the `SETUP_EDITOR` editor block. 

Would a middle ground here be to not split between action generators and action functions and just have `post-actions`, `editor-actions`, and `block-actions`?  That might also be useful for aiding future refactors where some (or all) of the block specific actions are moved elsewhere.

nerrad replied:
Keep in mind, I'm not only considering the line count for a combined actions file, but also the line count that would be exponentially larger for the corresponding tests file.  Currently combining `actions/test/post.js` and `actions/test/post-generators.js` into one file results in 759 lines.  As soon as a file gets over roughly 500 lines I find it starts to get unwieldy.  I imagine the line count for a combined actions file of editor, post, and block actions will start to go over 2000 easily.

nerrad replied:
> There's an interesting overlap between this discussion and some older ones like in #3030, where I think ultimately it was the introduction of `@wordpress/data` and its store-namespacing that served as the alternative to this problem of files becoming large and unwieldy, where the separation was by _stores_ (modules), not by separate files within a store.

To explore this further then, are you suggesting that we might want to consider applying some more granular namespacing here and split these up into more stores?  Of course, that could introduce some complications with deprecating the old actions etc but might be preferable over the file splitting.  So we could end up with something like `core/editor/posts`, `core/editor` (for general editor actions), and `core/editor/blocks` for block actions?

nerrad committed:
Add resolveSelect control and implement

nerrad commented on packages/editor/src/store/controls.js:
```diff
@@ -0,0 +1,78 @@
+			return registry.dispatch( reducerKey )[ actionName ]( ...args );
+		}
+	),
+	RESOLVE_SELECT: createRegistryControl(
```
It'd be nice if the registry exposed some sort of helper for determining if a selector has a resolver (i.e. `registry.getStore( reducerKey ).hasResolverFor( selectorName )`).  That way this code could be condensed to just having a `select` control that conditionally resolves if there is a resolver.

youknowriad replied:
Agreed, that some help from the registry here would be good, this could be addressed in a separate PR maybe?

nerrad committed:
refactor to condense code and update tests.

nerrad committed:
update docs

nerrad commented:
Okay this branch is up to date and I've refactored the work in here so all the post actions are condensed into `actions.js` (tests updated etc).  Tests pass locally, should be ready for review.  I am a bit surprised that functions prefixed with `__experimental` are showing up in the generated docs.  Not sure if that's something needing addressed or not.

aduth commented:
> Not sure if that's something needing addressed or not.

It does, but not here.

aduth commented on packages/editor/src/store/actions.js:
```diff
@@ -47,7 +64,6 @@ export function resetPost( post ) {
  * post has been received, by initialization or autosave.
  *
  * @param {Object} post Autosave post object.
- *
```
From which guideline are you asserting the documentation standard proposed here?

Elsewhere, the `@param` and `@return` are treated as separately grouped, within each group aligned on the name and descriptions, as documented here:

https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/#functions

There are multiple instances elsewhere in the pull request.

nerrad replied:
Huh, I've honestly never noticed this before.  Will correct.

nerrad replied:
So turns out my IDE (phpStorm) doesn't have much ability to customize the jsdocs template it creates.  So I've just never noticed that its defaults diverge so much from the wp core standards (and thus in my own projects I've just never thought anything of it).  I'll try to watch for this in the future.

aduth commented on packages/editor/src/store/actions.js:
```diff
@@ -57,6 +73,83 @@ export function resetAutosave( post ) {
+ * Optimistic action for indicating that the request post update has completed
+ * successfully.
+ *
+ * @param {Object} previousPost The previous post prior to update.
```
There's only one parameter to this function: an object. If the properties of the object are to be documented, they should be as prescribed by JSDoc and in the JavaScript Documentation Standards, e.g. `options.previousPost`

http://usejsdoc.org/tags-param.html#parameters-with-properties
https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/#functions

There are multiple instances elsewhere in the pull request.

nerrad replied:
> There's only one parameter to this function: an object.

Sorry I don't follow.  There's multiple arguments on this function not one.

nerrad replied:
Oh I follow now.  I think this was the doc block from the original function that I just copied over (but I could be wrong), I'll fix.

aduth commented on packages/editor/src/store/actions.js:
```diff
@@ -102,43 +195,276 @@ export function editPost( edits ) {
+ */
+export function* savePost( options = {} ) {
+	const isEditedPostSaveable = yield select(
+		MODULE_KEY,
```
I'm curious whether we should bother with this constant, considering there is precedent via hundreds of other components defined in the editor module referencing `'core/editor'` directly.

nerrad replied:
I think having the constant helps prevent against typos personally (however I do understand that reasoning can't be extrapolated to the store/module/reducer key being referenced outside of the package).

Another nit is that I see a variety of terms used to reference this argument in the codebase, sometimes its "reducer key", sometimes its "module key" and sometimes it is "store key".  I just went with `MODULE_KEY` here because that was the variable name given this originally when registering the store.

aduth replied:
> Another nit is that I see a variety of terms used to reference this argument in the codebase, sometimes its "reducer key", sometimes its "module key" and sometimes it is "store key". I just went with `MODULE_KEY` here because that was the variable name given this originally when registering the store.

Yes, it's been a lingering problem needing to be addressed. There wasn't already an issue, so I created one at #14150.

aduth commented on packages/editor/src/store/actions.js:
```diff
@@ -102,43 +195,276 @@ export function editPost( edits ) {
-export function autosave( options ) {
-	return savePost( { isAutosave: true, ...options } );
+export function* autosave( options ) {
+	yield* generatorActions.savePost( { isAutosave: true, ...options } );
```
Could this be simplified to just `yield* savePost( { /* ... */ } )j` ?

nerrad replied:
I used `generatorActions` here so I could mock the `savePost` action more easily.  It's possible I'm overlooking a way to more easily mock this in the tests or the mocking is unnecessary (we could just check that it yields a generator instead).

nerrad replied:
saw the subsequent comment about `generatorActions` and realize that ya I can just yield savePost directly I think here.

nerrad replied:
Won't work, need to keep this as is so it can be tested (see comment in other thread).

aduth commented on packages/editor/src/store/actions.js:
```diff
@@ -396,3 +722,12 @@ export const exitFormattedText = getBlockEditorAction( 'exitFormattedText' );
 export const updateBlockListSettings = getBlockEditorAction( 'updateBlockListSettings' );
 export const updateEditorSettings = getBlockEditorAction( 'updateEditorSettings' );
+
+// default export of generator actions.
```
I understand how you're using this for the unit tests, but it does not seem sensible to me that a default export from this file be a subset of generator actions.

Would it be possible instead to consider for the purposes of `jest.spyOn`, to use the `import * as actions from '../actions';` syntax?

nerrad replied:
doh, ya, I'll see if that works sufficiently for tests.

nerrad replied:
Turns out doing `import * as actions` and then `jest.spyOn( actions, 'savePost')` does not work.

aduth replied:
> Turns out doing `import * as actions` and then `jest.spyOn( actions, 'savePost')` does not work.

Why not?

nerrad replied:
No idea.  I tried it:

```js
import * as actions from '../actions';

// .... later in the test
describe( 'autosave()', () => {
		let savePostSpy;
		beforeAll( () => savePostSpy = jest.spyOn( actions, 'savePost' ) );
		afterAll( () => savePostSpy.mockRestore() );
		// autosave is mostly covered by `savePost` tests so just test the correct call
		it( 'calls savePost with the correct arguments', () => {
			const fulfillment = actions.autosave();
			fulfillment.next();
			expect( savePostSpy ).toHaveBeenCalled();
			expect( savePostSpy ).toHaveBeenCalledWith( { isAutosave: true } );
		} );
	} );
```

The test fails with the mock not getting called.  When I use `generatorActions` the mock works as expected.  I don't doubt there may be something I'm doing wrong but I'm guessing its because the mock isn't actually getting invoked by the autosave call.

aduth replied:
I sense it should be possible. The issue is likely that the spy creates a new reference which is different than the one called internal to the underlying `actions.js` file.

But I also wonder: Maybe rather than testing that it was called, we could test that the result of calling autosave would be the same as the result of calling `savePost` with the expected arguments?

Something like:

```js
it( 'calls savePost with the correct arguments', () => {
	expect( Array.from( autosave() ) ).toEqual( Array.from( savePost( { isAutosave: true } ) ) );
} );
```

nerrad replied:
Not sure how practical/useful that will be because they are both generators.

nerrad replied:
So one thing I _could_ do here is instead of yielding the `savePost` generator from `autosave`, I can just yield the `dispatch` control.  That way it's more easily testable and I can get rid of the `generatorActions` default export.

aduth commented on packages/editor/src/store/constants.js:
```diff
@@ -7,3 +7,22 @@
+export const POST_UPDATE_TRANSACTION_ID = 'post-update';
+export const SAVE_POST_NOTICE_ID = 'SAVE_POST_NOTICE_ID';
+export const TRASH_POST_NOTICE_ID = 'TRASH_POST_NOTICE_ID';
+export const PERMALINK_POSTNAME_REGEX = /%(?:postname|pagename)%/;
```
Where is this referenced? Was there an intention to consolidate constants from the `selectors.js` file here? Because that does not appear to be represented (the constants remain in `selectors.js` and are referenced as defined there).

nerrad replied:
hmm ya, intended to consolidate all constants into one file.  Makes it much easier to reference them (and helps with future refactors).  I'll review/fix.

aduth commented on packages/editor/src/store/controls.js:
```diff
@@ -1,17 +1,34 @@
 /**
  * WordPress dependencies
  */
+import { default as triggerFetch } from '@wordpress/api-fetch';
```
This could be simplified to:

```js
import triggerFetch from '@wordpress/api-fetch';
```

nerrad replied:
doh yup.

aduth commented on packages/editor/src/store/controls.js:
```diff
@@ -1,17 +1,34 @@
 import { createRegistryControl } from '@wordpress/data';
 
-/**
- * Dispatches an action.
```
We had a perfectly good DocBlock which was abolished?

nerrad replied:
hmm not intentional, maybe an artifact from resolving a merge conflict maybe.

nerrad replied:
Take note that this doc block isn't really accurate, so I suspect this was a merge conflict artifact (and likely the reason I resolved the way I did).  I suspect the doc block was on some other control action in here so to resolve, I'll just add doc blocks for all the actions :)

aduth commented on packages/editor/src/store/controls.js:
```diff
@@ -21,10 +38,41 @@ export function dispatch( storeKey, actionName, ...args ) {
 
-export default controls;
+				// trigger the selector (to trigger the resolver)
+				const result = getResult();
```
Minor: We could optimize this to avoid calling `getResult` until after we're certain it would be useful for the satisfied result of `hasFinished()`, i.e. call it inline with `return resolve( getResult() );` as done in the subscriber callback, or at least assigned as part of the `if` condition body.

nerrad replied:
hmm I just copied this directly from the `resolveSelector` function in `effects/utils` assuming via the comment that the selector needed to be called to trigger the resolution (otherwise `hasFinished()` would never return true).  The comment block made sense to me here.  Still feel it's unnecessary?

aduth replied:
Ah, I neglected to consider the preceding comment. It feels awkward to rely on a side-effect from a function named as a getter, but I suppose that's the nature of resolvers. It should be fine as-is.

aduth commented on packages/editor/src/store/utils/notice-builder.js:
```diff
@@ -0,0 +1,120 @@
+ *
+ * @param {Object} data Incoming data to build the arguments from.
+ * @return {Array}  Arguments for dispatch. An empty array signals no
+ * notification should be sent.
```
By convention, I'd recommend aligning all but the first lines with the start of the description from the previous line. This isn't documented in the JavaScript Documentation Standards, but is mentioned in passing in the PHP standards:

https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/php/#1-1-parameters-that-are-arrays

>(aligned with Description, if wraps to a new line)

nerrad replied:
The [WP standards](https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/#functions) you linked to earlier have descriptions aligned _per group_.  So for instance the @params have their descriptions aligned, and the @return is not necessarily aligned with the previous group (at least from the example given).  So that's what I followed in my pass through the doc blocks.

aduth replied:
> The [WP standards](https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/#functions) you linked to earlier have descriptions aligned _per group_. So for instance the @params have their descriptions aligned, and the @return is not necessarily aligned with the previous group (at least from the example given)

Yes, this sounds right. My point was toward the `@return` specifically aligning with itself. In your revisions, I think there are issues with mixed whitespace / tabbing.

![image](https://user-images.githubusercontent.com/1779930/53515134-4a262f80-3a97-11e9-9c5c-a68a1b40d631.png)

nerrad replied:
argh, looks fine in the IDE.  I *hate* when the IDE doesn't cooperate with documentation requirements.  I'll revisit but likely won't get to it today.

aduth reviewed (commented):
This seems like a solid refactor.

In case you're feeling particularly averse to merge conflicts, it might benefit you to squash most / all of your commits within the branch.

nerrad committed:
fix doc blocks

nerrad committed:
fix jsdocs for param object properties

nerrad committed:
simplify import statement

- this also initially experimented with just mocking `actions.savePost` for the autosave test but that didn’t work.  Needed to revert to the usage of a generatorActions default export for that.

nerrad committed:
utilize constants from constants file

nerrad committed:
remove unused constants (now in block-editor)

nerrad committed:
simplify import

nerrad committed:
add doc blocks for control actions

nerrad committed:
update docs

nerrad committed:
make sure test is importing used constant

nerrad commented:
Looks like tests are passing except for the undo related e2e tests that are currently failing everywhere else too.

nerrad committed:
fix doc block spacing issues

nerrad committed:
Swith constant name from `MODULE_KEY` to `STORE_KEY`

nerrad committed:
remove use of generatorActions default import ans switch to dispatch control for autosave

nerrad commented:
The [latest commits](https://github.com/WordPress/gutenberg/pull/13716/files/6d74643bcfea22604b41d797ad5b221a7db4c11b..4cc4015b16e21d896a492c954f29aefcf31db659)

- fix spacing issues in jsdoc blocks
- switch to use `STORE_KEY` instead of `MODULE_KEY` for the store name reference to contribute to the general renaming schema as initiated by #14150 
- switched `autosave()` generator action to use the dispatch control for dispatching the `savePost` action.  Thus I can remove the `generatorActions` default export and test that the autosave() is dispatching with the right args.

nerrad committed:
doc changes

aduth commented:
> * switched `autosave()` generator action to use the dispatch control for dispatching the `savePost` action.  Thus I can remove the `generatorActions` default export and test that the autosave() is dispatching with the right args.

While normally I'd be reluctant to alter the implementation for the sake of tests (in part the reason for the original review comment), it doesn't really strike me as any better or worse than the original implementation of the action creator to use the `dispatch` control this way. 👍

aduth commented on packages/editor/src/store/selectors.js:
```diff
@@ -27,18 +27,12 @@ import { createRegistrySelector } from '@wordpress/data';
- */
-export const POST_UPDATE_TRANSACTION_ID = 'post-update';
-const PERMALINK_POSTNAME_REGEX = /%(?:postname|pagename)%/;
-export const INSERTER_UTILITY_HIGH = 3;
```
I guess these were overlooked in the split for block editor? I verified they're not used in editor.

nerrad replied:
Either that or I resolved the merge incorrectly when bringing into this branch.

nerrad commented:
Ya I agree wrt to reluctance about altering implementation for the sake of tests.   But the benefit of this approach is that there isn't anything _additional_ added for the sake of easier testing (i.e. the previous behaviour).  This is a straight implementation that just happens to be more testable.

aduth commented:
The notice generated for a failed trash when offline is a bit unexpected. But the same exists on master, so ought be addressed separately. There might already be an issue.

![image](https://user-images.githubusercontent.com/1779930/53591411-ed437b80-3b61-11e9-99c5-912f1ed8ec1d.png)

aduth reviewed (approved):
Nice work here 👍

aduth committed:
Fix docs spacing

Co-Authored-By: nerrad <darren@roughsmootheng.in>
